### PR TITLE
Add SNI support to Factor

### PR DESCRIPTION
**Pull request by bpollack**
Untested on Windows, but I can guarantee it'll break there because I changed stack effects. That said, I'm unclear if it ever worked, so maybe we don't care.

I hate OpenSSL and it needs to die in a fire.

### CODE REVIEW

**Comment by bpollack** [comment]
Hold on this, it's super-close, but I'm doing a quick cleanup.

**Comment by bpollack** [comment]
Okay, theoretically good now.

**Comment by bpollack** [comment]
@erg This now supports `http-client` directly.

File: basis/io/sockets/secure/secure-docs.factor [hunk 1]
```diff
@@ -74,7 +74,7 @@ HELP: secure
 { $class-description "The class of secure socket addresses." } ;
 
 HELP: <secure>
-{ $values { "addrspec" "an address specifier" } { "secure" secure } }
+{ $values { "addrspec" "an address specifier" } { "host" "a string or " { $link f } } { "secure" secure } }
```
Comment (mrjbq7) [inline basis/io/sockets/secure/secure-docs.factor]: We have help markup ``{ $maybe string }`` for this case:

```factor
IN: scratchpad { $maybe string } print-element
a string or f
```

File: basis/io/sockets/secure/secure.factor [hunk 2]
```diff
@@ -44,16 +44,20 @@ HOOK: <secure-context> secure-socket-backend ( config -- context )
         with-disposal
     ] with-scope ; inline
 
-TUPLE: secure { addrspec read-only } ;
+TUPLE: secure
+    { addrspec read-only }
+    { hostname read-only } ;
 
-C: <secure> secure
+: <secure> ( addrspec host -- secure )
```
Comment (mrjbq7) [inline basis/io/sockets/secure/secure.factor]: Why can't this just be ``C: <secure> secure`` ?

File: basis/io/sockets/secure/secure.factor [hunk 3]
```diff
@@ -44,16 +44,20 @@ HOOK: <secure-context> secure-socket-backend ( config -- context )
...
+    secure boa ;
 
 M: secure present addrspec>> present " (secure)" append ;
 
 CONSULT: inet secure addrspec>> ;
 
 M: secure resolve-host ( secure -- seq )
-    addrspec>> resolve-host [ <secure> ] map ;
+    [ hostname>> ] [ addrspec>> ] bi
```
Comment (mrjbq7) [inline basis/io/sockets/secure/secure.factor]: Better without the swaps?
```factor
[ addrspec>> resolve-host ] [ hostname>> ] bi
[ <secure> ] curry map
```

File: basis/io/sockets/secure/unix/unix.factor [hunk 4]
```diff
@@ -14,10 +14,10 @@ M: ssl-handle handle-fd file>> handle-fd ;
 
 M: unix socket-handle fd>> ;
 
-M: secure ((client)) ( addrspec -- handle )
-    addrspec>> ((client)) <ssl-socket> ;
+M: secure ((client)) ( secure -- handle )
+    [ hostname>> ] [ addrspec>> ] bi ((client)) <ssl-socket> ;
```
Comment (mrjbq7) [inline basis/io/sockets/secure/unix/unix.factor]: It's weird that some words are ``( hostname addrspec -- )`` and others are ``( addrspec hostname -- )``...?

**Comment by mrjbq7** [comment]
Some minor comments, if you want to look it over and then tell me when ready to merge.

File: basis/io/sockets/secure/secure-docs.factor (hunk 1 above)
Comment (bpollack) [inline basis/io/sockets/secure/secure-docs.factor]: Fixed in a second.

File: basis/io/sockets/secure/secure.factor (hunk 2 above)
Comment (bpollack) [inline basis/io/sockets/secure/secure.factor]: It's a leftover from when I (temporarily) introduced a `<secure-host>` variant to aid in transitioning over to requiring the hostname everywhere. Fixed.

File: basis/io/sockets/secure/secure.factor (hunk 3 above)
Comment (bpollack) [inline basis/io/sockets/secure/secure.factor]: Good point, fixed.

File: basis/io/sockets/secure/unix/unix.factor (hunk 4 above)
Comment (bpollack) [inline basis/io/sockets/secure/unix/unix.factor]: Good catch. I realized halfway through that `( addrspec hostname -- ... )` is almost always the better stack ordering, but clearly didn't patch everywhere. Fixed.

**Comment by bpollack** [comment]
@mrjbq7 Thanks for the suggestions! I believe I have now implemented all of them, and we should be good to merge. If you want to test BTW, sni.velox.ch provides a quick test (compare that v. alice.sni.velox.ch or bob.sni.velox.ch).

**Comment by mrjbq7** [comment]
Great! Should we add a test case using that server?

**Comment by bpollack** [comment]
Sure. You want me to do that at a fairly raw level directly in `io.sockets.secure`, or put the actual test in `urls.secure` (which probably isn't where it belongs, but is a much more convenient place to stick the testing code).

**Comment by mrjbq7** [comment]
Either is fine, simpler is better :-)

**Comment by jonenst** [comment]
Do we have other tests that require an internet access and depend on public services ? Maybe add some kind of mechanism like platforms.txt to disable them ?

**Comment by bpollack** [comment]
@jonenst Well, the ones in `http.client` already do, and that's also the most convenient place to stick the SNI test, so that's what I did. @mrjbq7, if that's acceptable, I think that'll do it.

**Comment by erg** [comment]
We have a test that logs into a gmail account and moves folders around and
deletes them.

It would be nice to split up the monolithic tests concept into a super fast
test suite vs tests that require auth/network vs tests that take forever
(deploy).

**Comment by mrjbq7** [comment]
Merged, thanks!

**Pull request closed without merging**